internal/services/ingestor/buffered.go

```diff
@@ -0,0 +1,131 @@
+package ingestor
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hatchet-dev/hatchet/pkg/repository"
+	"github.com/hatchet-dev/hatchet/pkg/repository/prisma/dbsqlc"
+)
+
+type IngestBuf struct {
+	maxCapacity        int           // max number of events to hold in buffer before we flush
+	flushPeriod        time.Duration // max time to hold events in buffer before we flush
+	eventOpsChan       chan *eventBuffWrapper
+	lastFlush          time.Time
+	internalArr        []*eventBuffWrapper
+	sizeOfData         int // size of data in buffer
+	maxDataSizeInQueue int // max number of bytes to hold in buffer before we flush
+	BulkCreateFunc     func(ctx context.Context, opts *repository.BulkCreateEventSharedTenantOpts) (*repository.BulkCreateEventResult, error)
+}
+
+func NewIngestBuffer(maxCapacity int, flushPeriod time.Duration, maxDataSizeInQueue int, bulkCreateFunc func(ctx context.Context, opts *repository.BulkCreateEventSharedTenantOpts) (*repository.BulkCreateEventResult, error)) *IngestBuf {
+	return &IngestBuf{
+		maxCapacity:        maxCapacity,
+		flushPeriod:        flushPeriod,
+		eventOpsChan:       make(chan *eventBuffWrapper, maxCapacity*2),
+		lastFlush:          time.Now(),
+		internalArr:        make([]*eventBuffWrapper, 0),
+		sizeOfData:         0,
+		maxDataSizeInQueue: maxDataSizeInQueue,
+		BulkCreateFunc:     bulkCreateFunc,
+	}
+}
+
+type eventBuffWrapper struct {
+	eventOps *repository.CreateEventOpts
+	doneChan chan *flushResponse
+}
+
+func (b *IngestBuf) buffEventWorker(ctx context.Context) {
+	for {
+		select {
+		case e := <-b.eventOpsChan:
+			b.internalArr = append(b.internalArr, e)
+			b.sizeOfData += len(e.eventOps.Data)
+			b.sizeOfData += len(e.eventOps.AdditionalMetadata)
+
+			if len(b.internalArr) >= b.maxCapacity {
+				b.flush()
+			}
+			if b.sizeOfData >= b.maxDataSizeInQueue {
+				b.flush()
+			}
+
+		case <-time.After(time.Until(b.lastFlush.Add(b.flushPeriod))):
+			if len(b.internalArr) > 0 {
+				b.flush()
+			} else {
+				b.lastFlush = time.Now()
```

Comment:
I see that we set `lastFlush` here, should we be setting `lastFlush` when we hit the `case e := <-b.eventOpsChan` case as well?

Reply:
the case where we read from the channel should just buffer. We are setting lastFlush here because we are basically doing a no-op flush and otherwise we'd trigger again too soon in the after above as soon as we get a single item.